[WIP] add style transfer task with pystiche

pmeier opened the pull request:
## What does this PR do?

Add a style transfer task using [`pystiche`](https://pystiche.org) as backend.

Note: Change codeblock to test-code when 0.7.2 is out.  

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements) pystiche/pystiche#484
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

pep8speaks commented:
Hello @pmeier! Thanks for opening this PR. 









* In the file [`flash/vision/style_transfer/model.py`](https://github.com/PyTorchLightning/lightning-flash/blob/413a53021feb99be5b9d095b4b1f3cb1dfd3160a/flash/vision/style_transfer/model.py):

> [Line 90:49](https://github.com/PyTorchLightning/lightning-flash/blob/413a53021feb99be5b9d095b4b1f3cb1dfd3160a/flash/vision/style_transfer/model.py#L90): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`flash_examples/predict/style_transfer.py`](https://github.com/PyTorchLightning/lightning-flash/blob/413a53021feb99be5b9d095b4b1f3cb1dfd3160a/flash_examples/predict/style_transfer.py):

> [Line 47:41](https://github.com/PyTorchLightning/lightning-flash/blob/413a53021feb99be5b9d095b4b1f3cb1dfd3160a/flash_examples/predict/style_transfer.py#L47): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

pmeier commented on flash/vision/style_transfer/data.py:
```diff
@@ -0,0 +1 @@
+# TODO
```
I need an image dataset preferably unsupervised here. The original model was trained on the COCO 2014 train split.

edgarriba replied:
@pmeier consider to look at https://github.com/PyTorchLightning/lightning-flash/pull/256 since the api for loading data will change quite a bit

pmeier commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,90 @@
+        return output + input
+
+
+class Transformer(nn.Module):
```
This is the model that should be trained. It has no canonical name. This is compatible with the [PyTorch example](https://github.com/pytorch/examples/tree/master/fast_neural_style) which in turn follows [the initial paper](https://link.springer.com/content/pdf/10.1007%2F978-3-319-46475-6_43.pdf).

edgarriba replied:
as @tchaton mentioned, this model should be something that should be imported from `pystiche` and integrated it to the template task example that he provided

pmeier replied:
I'll leave it here for now. I need to push a release that includes this in `pystiche`. We shouldn't merge this PR until this is done.

edgarriba replied:
ok - then please mark it as Draft

pmeier replied:
Is it possible to review draft PRs?

edgarriba replied:
yes

pmeier commented on flash_examples/predict/style_transfer.py:
```diff
@@ -0,0 +1,47 @@
+    print("Please, run `pip install pystiche`")
+    sys.exit(0)
+
+multi_layer_encoder = enc.vgg16_multi_layer_encoder()
+
+content_layer = "relu2_2"
+content_encoder = multi_layer_encoder.extract_encoder(content_layer)
+content_weight = 1e5
+content_loss = ops.FeatureReconstructionOperator(
+    content_encoder, score_weight=content_weight
+)
+
+
+class GramOperator(ops.GramOperator):
+    def enc_to_repr(self, enc: torch.Tensor) -> torch.Tensor:
+        repr = super().enc_to_repr(enc)
+        num_channels = repr.size()[1]
+        return repr / num_channels
+
+
+style_layers = ("relu1_2", "relu2_2", "relu3_3", "relu4_3")
+style_weight = 1e10
+style_loss = ops.MultiLayerEncodingOperator(
+    multi_layer_encoder,
+    style_layers,
+    lambda encoder, layer_weight: GramOperator(encoder, score_weight=layer_weight),
+    layer_weights="sum",
+    score_weight=style_weight,
+)
+
+# TODO: this needs to be moved to the device to be trained on
+# TODO: we need to register a style image here
+perceptual_loss = loss.PerceptualLoss(content_loss, style_loss)
```
I'm not sure if this should go here or not. The perceptual loss together with the model are the core parts. So if the user wants to change things it is most likely one of the two.

pmeier commented on flash_examples/predict/style_transfer.py:
```diff
@@ -0,0 +1,47 @@
+)
+
+
+class GramOperator(ops.GramOperator):
+    def enc_to_repr(self, enc: torch.Tensor) -> torch.Tensor:
+        repr = super().enc_to_repr(enc)
+        num_channels = repr.size()[1]
+        return repr / num_channels
```
This change is only done to be compatible with the PyTorch example / paper. They both implement a double normalization by the number of channels which is most likely a bug. We could remove it here, but we would need to tune the hyperparameters then.

edgarriba replied:
from flash perspective I believe this is something that should happen internally in your library. /cc @tchaton

pmeier reviewed (commented):
This is just the very preliminary state. I've hit several roadblocks that need to be resolved:

- Models for Neural Style Transfer are trained in an unsupervised manner. Thus, I simply need a dataset that can supply images without labels / annotations.
- Following from the point above, there is no train / val / test split. The process is done after the training. If there is something like a validation / test it is performed manually by trying a few examples. There is no objective way to put a number on the quality of the stylization.
- The models used as transformer are not named. I've seen that the model is usually loaded by their name, which is thus not possible. We could fall back to a author / year combination of the paper the architecture was published. 

I'll fix the linting errors and update the documentation, tests, and the changelog when the main part is resolved.

tchaton commented on flash/vision/__init__.py:
```diff
@@ -2,3 +2,4 @@
 from flash.vision.classification import ImageClassificationData, ImageClassificationPreprocess, ImageClassifier
 from flash.vision.detection import ObjectDetectionData, ObjectDetector
 from flash.vision.embedding import ImageEmbedder
+from .style_transfer import *
```
use direct import everywhere :)

tchaton commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,90 @@
+__all__ = ["Transformer"]
+
+
+class Interpolate(nn.Module):
```
Could this model be imported directly from `pystiche`. The goal of Flash is to rely on other libraries to enable tasks, not implement nn layers :)

edgarriba replied:
or use the `Resize`/`Rescale` module with those options that you implemented in `kornia` :)

tchaton commented on flash_examples/predict/style_transfer.py:
```diff
@@ -0,0 +1,47 @@
+
+style_layers = ("relu1_2", "relu2_2", "relu3_3", "relu4_3")
+style_weight = 1e10
+style_loss = ops.MultiLayerEncodingOperator(
```
Everything should be included in a Flash Task.

tchaton commented:
Hey @pmeier,

Awesome you started.

```
class StyleTransfer(Task):

    models: FlashRegistry = STYLE_TRANSFER_MODELS

    def __init__(
        self,

        content_image: Union[Image.PIL, str, np.ndarray],
        style_loss: Optional[Callable] = None,
        content_loss: Optional[Callable] = None,
        perceptual_loss: Optional[Callable] = None,

        model: Union[str, Tuple[nn.Module, int]] = "transformer",
        model_kwargs: Optional[Dict] = None,
        optimizer: Union[Type[torch.optim.Optimizer], torch.optim.Optimizer] = torch.optim.Adam,
        optimizer_kwargs: Optional[Dict[str, Any]] = None,
        scheduler: Optional[Union[Type[_LRScheduler], str, _LRScheduler]] = None,
        scheduler_kwargs: Optional[Dict[str, Any]] = None,
        metrics: Union[torchmetrics.Metric, Mapping, Sequence, None] = None,
        learning_rate: float = 1e-3,
        serializer: Optional[Union[Serializer, Mapping[str, Serializer]]] = None,
    ):

        if perceptual_loss is None:
            content_loss = content_loss or self.default_content_loss()
            style_loss = style_loss or self.default_style_transfer()
            perceptual_loss = loss.PerceptualLoss(content_loss, style_loss)

        if content_image is not None:
            perceptual_loss.set_content_image(content_image)

        self.perceptual_loss = perceptual_loss

        self.save_hyperparameters()

        if isinstance(model, tuple):
            model = model
        else:
            model = self.models.get(model)(pretrained=pretrained, **model_kwargs)


        super().__init__(
            model=model,
            loss_fn=perceptual_loss,
            optimizer=optimizer,
            optimizer_kwargs=optimizer_kwargs,
            scheduler=scheduler,
            scheduler_kwargs=scheduler_kwargs,
            metrics=metrics,
            learning_rate=learning_rate,
            serializer=serializer,
        )


    def default_content_loss(self):
        multi_layer_encoder = enc.vgg16_multi_layer_encoder()
        content_layer = "relu2_2"
        content_encoder = multi_layer_encoder.extract_encoder(content_layer)
        content_weight = 1e5
        return = ops.FeatureReconstructionOperator(
            content_encoder, score_weight=content_weight
        )


    def default_style_transfer(self):
        class GramOperator(ops.GramOperator):
            def enc_to_repr(self, enc: torch.Tensor) -> torch.Tensor:
                repr = super().enc_to_repr(enc)
                num_channels = repr.size()[1]
                return repr / num_channels


        style_layers = ("relu1_2", "relu2_2", "relu3_3", "relu4_3")
        style_weight = 1e10
        return ops.MultiLayerEncodingOperator(
            multi_layer_encoder,
            style_layers,
            lambda encoder, layer_weight: GramOperator(encoder, score_weight=layer_weight),
            layer_weights="sum",
            score_weight=style_weight,
        )

    def forward(self, x):
        # not sure about this part
        self.model(x)
        return self.perceptual_loss(x)



# in finetuning.

content_image = ...
dm = StyleDataModule.from_folder(...)
model = StyleTransfer(content_image=content_image)
trainer = Trainer(...)
trainer.fit(model, dm)
```

pmeier commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,185 @@
+        self.perceptual_loss = loss.PerceptualLoss(content_loss, style_loss)
+        self.perceptual_loss.set_style_image(style_image)
+
+        self.save_hyperparameters()
```
Couldn't find documentation for this. Is this needed? If yes, what does it do?

edgarriba replied:
@tchaton @ethanwharris  is this somethings that we could automate inside the main `Task` class ?

pmeier commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,185 @@
+
+    def __init__(
+        self,
+        style_image: torch.Tensor,
```
`pystiche` has built-in support for reading images from path. If you want support for `PIL.Image.Image` and `np.ndarray` I need some functionality to transform them into a tensor. `torchvision` has [`transforms.ToTensor`](https://pytorch.org/vision/stable/transforms.html#torchvision.transforms.ToTensor) for that.

edgarriba replied:
if you already support torchvision internally - you could somehow expose or implement that internally

ethanwharris replied:
You should be able to configure the to tensor transform in the default_transforms in the preprocess

pmeier commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,185 @@
+        optimizer_kwargs: Optional[Dict[str, Any]] = None,
+        scheduler: Optional[Union[Type[_LRScheduler], str, _LRScheduler]] = None,
+        scheduler_kwargs: Optional[Dict[str, Any]] = None,
+        metrics: Union[torchmetrics.Metric, Mapping, Sequence, None] = None,
```
I see no point in accepting this, since there are no metrics for style transfer.

edgarriba replied:
I would keep it for consistency with others task and maybe raise an error with a meaningful message. We don't know in future if anyone will come with goof metric for this task

edgarriba replied:
Besides, what do you look at when you train style transfer - just the loss ?

ethanwharris replied:
Yeah, I would remove it then just add it back if one day there is a metric

pmeier commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,185 @@
+        scheduler_kwargs: Optional[Dict[str, Any]] = None,
+        metrics: Union[torchmetrics.Metric, Mapping, Sequence, None] = None,
+        learning_rate: float = 1e-3,
+        serializer: Optional[Union[Serializer, Mapping[str, Serializer]]] = None,
```
I didn't find any documentation on this. At least it is not explained in the docstring of `flash.Task`. I'm guessing this can stay, but IDK.

edgarriba replied:
this is a relatively new flash component that helps to retrieve and serialize the expected results for a task. E.g. apply correct mapping to classification tasks. My guess is that in general image generation this just needs to return/save the final image. /cc @ethanwharris

ethanwharris replied:
Yeah, you may want a serializer that converts the style transferred tensor back into a PIL image to then be viewed.

pmeier commented on flash_examples/predict/style_transfer.py:
```diff
@@ -0,0 +1,23 @@
+
+download_data("http://images.cocodataset.org/zips/train2014.zip", "data")
+
+data_module = ImageUnsupervisedData.from_folder("data")
```
I need something like a `ImageUnsupervised` class. If that is not possible, we could also use any other data loading class that returns all images. In that case, I need to know how to discard the class labels or annotations.

edgarriba replied:
what are usually the inputs/labels for style transfer - pair of images ? We are about to introduce a similar data loading api for semantic segmentation that might be extend to other image generation tasks. /cc @ethanwharris 

@pmeier https://github.com/PyTorchLightning/lightning-flash/pull/239/files#diff-fff239f44797f0092593a14bf1af9a4b97d1959cba4a6a759eadbf126a697fc5R41

pmeier replied:
> what are usually the inputs/labels for style transfer - pair of images ?

Style transfer is unsupervised, so there are not pairs. I just need the images.

pmeier commented on flash_examples/predict/style_transfer.py:
```diff
@@ -0,0 +1,23 @@
+
+model = StyleTransfer(style_image)
+
+trainer = flash.Trainer(max_epochs=2)
```
I also need `batch_size=4` but IDK where to put it. It doesn't seem to belong to the trainer.

edgarriba replied:
Flash Trainer sub-classes PL-Trainer which has `batch_size` as a member variable

pmeier replied:
There is no mention of a `batch_size` in the [docs](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.trainer.trainer.html#trainer).

ethanwharris replied:
`batch_size` is passed to the `DataModule` in the `from_*` method, not the trainer. So you should be able to do: `from_folders(..., batch_size=4)` :smiley:

pmeier reviewed (commented):
Apart from my comments / questions below, I'm wondering whether this example should be in `predict` or `finetune`.

ethanwharris commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,210 @@
+        self.perceptual_loss.set_content_image(content_image)
+        return self.model(content_image)
+
+    def validation_step(self, batch: Any, batch_idx: int) -> NoReturn:
```
Should also override `train_step` and `predict_step` here to get the `DefaultDataKeys.INPUT` from the batch. Example here: https://github.com/PyTorchLightning/lightning-flash/blob/d5ca5e526ff1cb3bad23a48398c4634b7fa37724/flash/vision/classification/model.py#L110

ethanwharris reviewed (commented):
It's getting there, just a couple of comments :smiley:

pmeier commented:
Training is running locally at the moment. Afterwards I'll update the documentation and the changelog and add tests.

pmeier commented on flash_examples/finetuning/style_transfer.py:
```diff
@@ -0,0 +1,25 @@
+from flash.utils.imports import _PYSTICHE_AVAILABLE
+from flash.vision.style_transfer import StyleTransfer, StyleTransferData
+
+if _PYSTICHE_AVAILABLE:
```
If `pystiche` is not available, we'll never get to this, since this PR makes `flash` dependent on `pystiche` at import. Should this not be the case?

tchaton replied:
We will refactor requirement, but Flash will depend on  pystiche only if user do `pip install flash['style-transfer']` and we need to make sure the code doesn't break your library isn't installed.

pmeier replied:
But until this happens, this check is moot. Even after the refactor it is probably useless, since importing `StyleTransfer` depends on `pystiche`, so we need to do the guarding there.

pmeier replied:
After #297 `pystiche` is now an optional dependency.

tchaton commented on flash/vision/style_transfer/data.py:
```diff
@@ -0,0 +1,102 @@
+__all__ = ["StyleTransferPreprocess", "StyleTransferData"]
+
+
+def _apply_to_input(default_transforms_fn) -> Callable[..., Dict[str, ApplyToKeys]]:
```
Maybe worth to move this to transform.py as it could be reused.

tchaton commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,227 @@
+
+        if model is None:
+            # TODO: import this from pystiche
+            model = Transformer()
```
`pystiche` should a hub of model (pretrained possibly) and make them available through the registry.

pmeier replied:
Yes, it should, but it probably won't in the next time. I only have limited time / resources to maintain `pystiche` so it is only one model for now.

tchaton commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,227 @@
+            model = Transformer()
+
+        if multi_layer_encoder is None:
+            multi_layer_encoder = self.default_multi_layer_encoder()
```
I guess the default could also come with the registry.

pmeier replied:
Not sure what you mean. Could you elaborate?

tchaton commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,227 @@
+        )
+
+        self.perceptual_loss = perceptual_loss
+        self.perceptual_loss.set_style_image(style_image)
```
Is it possible to set multiple style_images and to randomly sample from them ?

pmeier replied:
There are approaches like that, but they go beyond a simple entry example.

tchaton commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,227 @@
+        self, multi_layer_encoder: Optional[enc.MultiLayerEncoder] = None
+    ) -> ops.MultiLayerEncodingOperator:
+
+        class GramOperator(ops.GramOperator):
```
Should come from `pystiche`

pmeier replied:
Can't. I'm going for compatibility here with the PyTorch example / paper implementation and they both do this. This is bug / oversight in their implementation.

pmeier replied:
I've added an explanation through a comment.

tchaton replied:
Make sense :)

tchaton commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,227 @@
+
+        style_layers = ("relu1_2", "relu2_2", "relu3_3", "relu4_3")
+        style_weight = 1e10
+        return ops.MultiLayerEncodingOperator(
```
Worth to expose those elements to the task as `style_loss_kwargs`

tchaton replied:
Same for all configurable elements.

pmeier replied:
How is that easier than defining the loss outside of the task and pass it directly?

tchaton commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,227 @@
+        output_image = self(batch)
+        loss = self.perceptual_loss(output_image).total()
+
+        logs = dict(perceptual_loss=loss)
```
Do you have any metrics ? Also better to use self.log to log metrics. Returning a dictionary is -/+ depreciated.

pmeier replied:
> Do you have any metrics ?

There are none.

>  Also better to use self.log to log metrics. Returning a dictionary is -/+ depreciated.

Really? I mean I basically just did a simpler version of `super().step()` here, which also returns a dictionary.

tchaton commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,227 @@
+        return super().training_step(batch[DefaultDataKeys.INPUT], batch_idx)
+
+    def validation_step(self, batch: Any, batch_idx: int) -> NoReturn:
+        raise_not_supported("validation")
```
Not sure entirely why validation isn't supported. 
Also, might be better to set `self.validation_step = None` in the init function.

pmeier replied:
Style transfer is unsupervised so there is no reason for a validation. What labels what you validate against?

tchaton commented on flash_examples/finetuning/style_transfer.py:
```diff
@@ -0,0 +1,25 @@
+
+data_module = StyleTransferData.from_folders(train_folder="data", batch_size=4)
+
+style_image = pystiche.demo.images()["paint"].read(size=256)
```
Can we replace this by an image from internet, like 

![](https://www.kazoart.com/blog/wp-content/uploads/2017/03/van-gogh-la-nuit-etoilee-1.jpg).

style_image could actually be url to style_transfer.

pmeier replied:
Sure, just tell me which image you want to use. Not sure about licensing though. The images I used are licensed very permissive.

edgarriba replied:
as soon we reference well should be fine. Consider also adding a link to here in the docs: https://paperswithcode.com/task/style-transfer. Papers with code will be reference soon also to submit results.

pmeier replied:
> Consider also adding a link to here in the docs: https://paperswithcode.com/task/style-transfer. Papers with code will be reference soon also to submit results.

The link is doing image-optimization whereas we are implementing model-optimization here.

pmeier replied:
@tchaton I need a reference image to check if the training works well with the default parameters.

tchaton replied:
Oh I see. Makes sense.

tchaton reviewed (commented):
Lot of great progress. Getting there.

pmeier commented on flash/vision/style_transfer/data.py:
```diff
@@ -0,0 +1,102 @@
+
+    @property
+    def default_val_transforms(self) -> None:
+        # Style transfer doesn't support a validation phase, so we return nothing here
```
IMO it will always be like this. The problem is that in order to validate or test, you need an objective metric to measure the model performance. With style transfer this metric needs to answer the question "how well is this image stylized given this were the original content and style". This is impossible to do in the current state of the art and I wager a guess it will stay that way.

edgarriba replied:
sounds good - would be great to have that piece of information somewhere in the code or docs, or in both :)

pmeier commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,227 @@
+        style_image: Union[str, torch.Tensor],
+        model: Optional[nn.Module] = None,
+        multi_layer_encoder: Optional[enc.MultiLayerEncoder] = None,
+        content_loss: Optional[Union[ops.ComparisonOperator, ops.OperatorContainer]] = None,
```
This goes against what @tchaton proposed in https://github.com/PyTorchLightning/lightning-flash/pull/262#issuecomment-833300669. I'm happy to do either, but you (plural) need to decide which variant you would like to have.

edgarriba replied:
ops, missed that detail. What do you think @tchaton  ? My proposal goes in the direction to have a more generic api across tasks..

tchaton commented on flash/vision/style_transfer/backbone.py:
```diff
@@ -0,0 +1,23 @@
+
+STYLE_TRANSFER_BACKBONES = FlashRegistry("backbones")
+
+for mle_fn in dir(enc):
```
Neat !

tchaton commented on flash/vision/style_transfer/data.py:
```diff
@@ -0,0 +1,111 @@
+
+    @property
+    @_apply_to_input
+    def default_train_transforms(self) -> Dict[str, Callable]:
```
@ethanwharris @pmeier We made a new addition. This code can be simplified.

Now, {}_default_transforms is a Preprocess hook.

You can also do:
```
def default_transforms(self):
    if not self.training or self.predicting: return None 
    if self.training:
        dict(
            to_tensor_transform=torchvision.transforms.ToTensor(),
            per_sample_transform_on_device=nn.Sequential(
                transforms.Resize(min(self.image_size)),
                transforms.CenterCrop(self.image_size),
            ),
        )
    else:
         dict(
            to_tensor_transform=torchvision.transforms.ToTensor(),
            per_sample_transform_on_device=transforms.Resize(min(self.image_size)),
        )

```

pmeier replied:
I've encountered a problem with this: since I'm inheriting `ImageClassificationPreprocess` the `default_transforms_train` takes a higher priority. I think this is not intended. It happens because "prefix" (should be postfix BTW) is an empty string

https://github.com/PyTorchLightning/lightning-flash/blob/8eef52fa9b72befab339e388ce862e09f5f28e89/flash/core/data/data_pipeline.py#L183

but the check if for overridden methods checks for `None` explicitly:

https://github.com/PyTorchLightning/lightning-flash/blob/8eef52fa9b72befab339e388ce862e09f5f28e89/flash/core/data/data_pipeline.py#L131

Thus, `current_method_name` becomes `_default_transforms` which is indeed not overridden.

pmeier replied:
Fixed in 21d0817.

tchaton commented on flash/vision/style_transfer/model.py:
```diff
@@ -0,0 +1,122 @@
+        content_layer: str = "relu2_2",
+        content_weight: float = 1e5,
+        style_layers: Sequence[str] = ("relu1_2", "relu2_2", "relu3_3", "relu4_3"),
+        style_weight: float = 1e10,
```
Looks really neat !

pmeier commented on flash/core/utilities/imports.py:
```diff
@@ -74,6 +74,7 @@ def _compare_version(package: str, op, version) -> bool:
 _PYTORCHVIDEO_AVAILABLE = _module_available("pytorchvideo")
 _MATPLOTLIB_AVAILABLE = _module_available("matplotlib")
 _TRANSFORMERS_AVAILABLE = _module_available("transformers")
+_PYSTICHE_AVAILABLE = _module_available("pystiche")
```
Should this also be added to `_IMAGE_AVAILABLE`?

tchaton replied:
I think not doing so for now is better. I need to think about a way to segregate some deps

pmeier commented on flash/core/data/data_pipeline.py:
```diff
@@ -180,7 +180,7 @@ def _resolve_function_hierarchy(
         if object_type is None:
             object_type = Preprocess
 
-        prefixes = ['']
```
See https://github.com/PyTorchLightning/lightning-flash/pull/262#discussion_r633089549.

pmeier commented:
@tchaton @ethanwharris @edgarriba I need some input:

- Is the documentation skeleton that I added sufficient?
- What exactly should be tested? I've added a disabled finetuning testing, since it will train for two epochs otherwise.

pmeier commented on requirements/datatype_image.txt:
```diff
@@ -5,3 +5,5 @@ Pillow>=7.2
 kornia>=0.5.1
 matplotlib
 pycocotools>=2.0.2 ; python_version >= "3.7"
+git+https://github.com/pystiche/pystiche@releases/v0.7.2
+# pystiche>=0.7.2
```
I'll push a new release as soon as we are sure nothing needs to be fixed / added to `pystiche` anymore.

tchaton commented on docs/source/reference/style_transfer.rst:
```diff
@@ -0,0 +1,35 @@
+********
+
+TODO
+
```
Mind describing the task ?

edgarriba replied:
link to here would be nice too: https://paperswithcode.com/task/style-transfer

pmeier replied:
@edgarriba see https://github.com/PyTorchLightning/lightning-flash/pull/262#discussion_r631697099

tchaton commented on docs/source/reference/style_transfer.rst:
```diff
@@ -0,0 +1,35 @@
+Fit
+***
+
+TODO
```
Same here ?

tchaton commented on flash/image/style_transfer/data.py:
```diff
@@ -0,0 +1,103 @@
+from flash.core.data.transforms import ApplyToKeys
+from flash.image.classification import ImageClassificationData, ImageClassificationPreprocess
+
+from ._utils import raise_not_supported
```
Use full path

tchaton commented on tests/examples/test_scripts.py:
```diff
@@ -94,6 +94,11 @@ def run_test(filepath):
             "translation.py",
             marks=pytest.mark.skipif(not _TEXT_AVAILABLE, reason="text libraries aren't installed")
         ),
+        # pytest.param(
```
Could you add some tests ?

tchaton replied:
By default, it runs fast_dev_run=True. I guess couple of tests to get the coverage up.
Plus, need to make sure the imports don t break if your lib is missing

pmeier commented on flash/core/data/data_pipeline.py:
```diff
@@ -190,9 +191,11 @@ def _resolve_function_hierarchy(
         elif stage == RunningStage.PREDICTING:
             prefixes += ['predict']
 
+        prefixes += [None]
```
This would mean that `default_transforms` has a lower priority as `default_transforms_train`. IMO this will not fix https://github.com/PyTorchLightning/lightning-flash/pull/262#discussion_r633089549. If I'm not mistaken this will pull the default train transform which is unwanted

tchaton replied:
Removed ImageClassificationPrepreprocess, so we don't have this issue.

tchaton replied:
Yes, `train_defaut_transforms` should have a higher priority than `default_transforms`.

pmeier commented on flash/image/style_transfer/backbone.py:
```diff
@@ -0,0 +1,24 @@
+        if "multi_layer_encoder" not in mle_fn:
+            continue
+
+        name = mle_fn.split("_")[0]
```
This will result in the same name for `vgg19` and `vgg19_bn`. Why not keep the regex matching?

tchaton replied:
Reverted back.

pmeier commented on flash/image/style_transfer/backbone.py:
```diff
@@ -0,0 +1,24 @@
+
+        STYLE_TRANSFER_BACKBONES(
+            fn=lambda: (getattr(enc, mle_fn)(), None),
+            name=mle_fn.split("_")[0],
```
```suggestion
            name=name,
```

tchaton replied:
Reverted back.

pmeier commented on flash_examples/predict/semantic_segmentation.py:
```diff
@@ -24,7 +24,9 @@
 
 # 2. Load the model from a checkpoint
-model = SemanticSegmentation.load_from_checkpoint("semantic_segmentation_model.pt")
+model = SemanticSegmentation.load_from_checkpoint(
```
Not sure what this has to do with the PR?

tchaton replied:
It has a tiny bug, nothing major

pmeier commented on requirements/datatype_image_style_transfer.txt:
```diff
@@ -0,0 +1,4 @@
+torchvision
+Pillow>=7.2
+kornia>=0.5.1
+pystiche
```
```suggestion
pystiche>=0.7.2
```

tchaton replied:
Done !

pmeier commented on tests/examples/test_scripts.py:
```diff
@@ -94,6 +97,11 @@ def run_test(filepath):
             "translation.py",
             marks=pytest.mark.skipif(not _TEXT_AVAILABLE, reason="text libraries aren't installed")
         ),
+        pytest.param(
+            "finetuning",
+            "style_transfer.py",
+            marks=pytest.mark.skipif(not _IMAGE_STLYE_TRANSFER, reason="pystiche is not installed")
+        ),
```
So this should run with every PR run? Each run will take ~2 hours.

tchaton replied:
It is running with `FLASH_TESTING=1` which maps to `fast_dev_run=True` within the trainer.

pmeier reviewed (commented):
Hey @tchaton, thanks for the commits. I have some comments below. Additionally, it looks like you have added a lot of changes that seemingly have nothing to do with this PR. Was that intentional?

edgarriba commented on flash/image/style_transfer/model.py:
```diff
@@ -0,0 +1,174 @@
+    from pystiche.image import read_image
+else:
+
+    class enc:
```
`CamelCase` class name

tchaton replied:
Following pystiche.

pmeier replied:
This just happens because these classes are actually modules.

edgarriba commented on flash/image/style_transfer/model.py:
```diff
@@ -0,0 +1,174 @@
+        content_weight: float = 1e5,
+        style_layers: Union[Sequence[str], str] = ("relu1_2", "relu2_2", "relu3_3", "relu4_3"),
+        style_weight: float = 1e10,
+        optimizer: Union[Type[torch.optim.Optimizer], torch.optim.Optimizer] = torch.optim.Adam,
```
in same tasks we are promoting `AdamW` since it has a better implementation of the original algorithm. Have you ever tried here ?

pmeier replied:
Nope. Until @tchaton changes this followed the official PyTorch example. Since we are deviating now, we could also use another optimizer.

edgarriba reviewed (approved):
LGTM

ethanwharris reviewed (approved):
Awesome! Small comment
